File: server/models/Product.js
const mongoose = require('mongoose');

const SourceSchema = new mongoose.Schema({
  retailer: {
    type: String,
    enum: ['flipkart', 'amazon', 'croma', 'reliance'],
    required: true
  },
  url: {
    type: String,
    required: true
  },
  currentPrice: {
    type: Number,
    required: true
  },
  mrp: {
    type: Number,
    default: 0
  },
  discountPct: {
    type: Number,
    default: 0
  },
  rating: {
    type: Number,
    default: 0
  },
  inStock: {
    type: Boolean,
    default: true
  },
  seller: {
    type: String,
    default: 'Verified Seller'
  },
  deliveryText: {
    type: String,
    default: 'Free Delivery'
  },
  lastCheckedAt: {
    type: Date,
    default: Date.now
  }
}, { _id: false });

const ProductSchema = new mongoose.Schema({
  name: {
    type: String,
    required: [true, 'Product name is required'],
    trim: true,
    index: 'text'
  },
  slug: {
    type: String,
    unique: true,
    lowercase: true,
    index: true
  },
  brand: {
    type: String,
    trim: true,
    index: true
  },
  category: {
    type: String,
    required: true,
    index: true
  },
  description: {
    type: String,
    default: ''
  },
  images: [{
    type: String
  }],
  thumbnail: {
    type: String,
    default: ''
  },
  specs: {
    type: Map,
    of: String,
    default: {}
  },
  sources: [SourceSchema],
  cheapestSource: {
    retailer: String,
    price: Number,
    savingsVsMrp: Number
  },
  allTimeLow: {
    price: Number,
    retailer: String,
    date: Date
  },
  allTimeHigh: {
    price: Number,
    retailer: String,
    date: Date
  }
}, {
  timestamps: true
});

// Middleware to calculate cheapest source before saving
ProductSchema.pre('save', function(next) {
  if (this.sources && this.sources.length > 0) {
    const validSources = this.sources.filter(s => s.inStock && s.currentPrice > 0);
    if (validSources.length > 0) {
      const cheapest = validSources.reduce((min, s) => s.currentPrice < min.currentPrice ? s : min, validSources[0]);
      const maxMrp = Math.max(...this.sources.map(s => s.mrp || s.currentPrice));
      this.cheapestSource = {
        retailer: cheapest.retailer,
        price: cheapest.currentPrice,
        savingsVsMrp: maxMrp > cheapest.currentPrice ? maxMrp - cheapest.currentPrice : 0
      };
    }
  }
  if (!this.thumbnail && this.images && this.images.length > 0) {
    this.thumbnail = this.images[0];
  }
  next();
});

module.exports = mongoose.model('Product', ProductSchema);

File: data-service/schemas/product.py
from pydantic import BaseModel, Field
from typing import List, Optional, Dict
from datetime import datetime

class RetailerSourceSchema(BaseModel):
    retailer: str = Field(..., description="flipkart, amazon, croma, or reliance")
    url: str
    currentPrice: float
    mrp: Optional[float] = 0.0
    discountPct: Optional[float] = 0.0
    rating: Optional[float] = 0.0
    inStock: bool = True
    seller: Optional[str] = "Verified Seller"
    deliveryText: Optional[str] = "Standard Delivery"
    lastCheckedAt: datetime = Field(default_factory=datetime.utcnow)

class CheapestSourceSchema(BaseModel):
    retailer: str
    price: float
    savingsVsMrp: float = 0.0

class NormalizedProductSchema(BaseModel):
    name: str
    slug: str
    brand: Optional[str] = ""
    category: str = "Electronics"
    description: Optional[str] = ""
    images: List[str] = []
    thumbnail: Optional[str] = ""
    specs: Dict[str, str] = {}
    sources: List[RetailerSourceSchema] = []
    cheapestSource: Optional[CheapestSourceSchema] = None

class ScrapeSearchRequest(BaseModel):
    query: str
    category: Optional[str] = "All"
    limit: Optional[int] = 10

class ScrapeUrlRequest(BaseModel):
    url: str
    retailer: Optional[str] = None

class PriceHistoryItem(BaseModel):
    productId: str
    retailer: str
    price: float
    checkedAt: datetime = Field(default_factory=datetime.utcnow)
